tracing/span.go

```diff
@@ -0,0 +1,141 @@
+package tracing
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Span is either an OpenTracing span or an OpenTelemetry span.
+// It could also be a noop span.
+type Span struct {
+	opentracingSpan opentracing.Span
+	otelSpan        trace.Span
+}
+
+// StartSpanFromContext starts a new span from the context using the parent tracing library.
+// If parent span is not sampled, it returns a noop span.
+func StartSpanFromContext(ctx context.Context, operation string, options ...SpanOption) (*Span, context.Context) {
+	otelSpan, opentracingSpan, sampled := SpanFromContext(ctx)
+	if !sampled {
+		return &Span{}, ctx
+	}
+
+	if opentracingSpan != nil {
+		var opentracingOptions []opentracing.StartSpanOption
+		for _, opt := range options {
+			opentracingOptions = append(opentracingOptions, opt.opentracingSpanOptions()...)
+		}
+		span, ctx := opentracing.StartSpanFromContext(ctx, operation, opentracingOptions...)
+		s := &Span{opentracingSpan: span}
+		for _, opt := range options {
+			opt.apply(s)
+		}
+		return s, ctx
+	}
+
+	if otelSpan != nil {
+		var otelOptions []trace.SpanStartOption
+		for _, opt := range options {
+			otelOptions = append(otelOptions, opt.otelSpanOptions()...)
+		}
+		ctx, span := otelSpan.TracerProvider().Tracer("dskit/tracing").Start(ctx, operation, otelOptions...)
+		s := &Span{otelSpan: span}
+		for _, opt := range options {
+			opt.apply(s)
+		}
+		return s, ctx
+	}
+	return &Span{}, ctx
```

Comment:
This is a different behavior compared to before the PR, since `opentracing.StartSpanFromContext` would create a real span if there was no parent: see https://github.com/opentracing/opentracing-go/blob/10b1cf09e00bdc84234b8c7a4b4d4e4afe64de87/gocontext.go#L63

Reply:
You're right, fixed here and added some tests: https://github.com/grafana/dskit/pull/681/commits/51ca0c0e327048a78dcd78bb56ae54b6cbfe054c